packages/ipfs-message-port-protocol/README.md

```diff
@@ -0,0 +1,214 @@
+
+## Wire protocol codecs
+
+Library provides encode / decode functions for types that are not supported by [structured cloning algorithm][] and therefore need to be encoded before posted over [message channel][] and decoded on the other end.
+
+All encoders take optional `transfer` array. If provided, encoder will add all `Transferable` fields of the given value so the they could be moved across threads without copying.
+
+### `CID`
+
+Codecs for [CID][] implementation in JavaScript.
+
+```js
+const { CID, encodeCID, decodeCID } = require('ipfs-message-port-protocol/src/cid')
+
+const cid = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
+
+const { port1, port2 } = new MessageChannel()
+
+// Will copy underyling memory
+port1.postMessage(encodeCID(cid))
+
+// Will transfer underlying memory (cid is corrupt on this thread)
+const transfer = []
+port1.postMessage(encodeCID(cid, transfer), transfer)
+
+// On the receiver thread
+port2.onmessage = ({data}) => {
+  const cid = decodeCID(data)
+  data instanceof CID // => true
+}
+```
+
+### Block
+
+Codecs for [IPLD Block][] implementation in JavaScript.
+
+```js
+const { Block, encodeBlock, decodeBlock } = require('ipfs-message-port-protocol/src/block')
+
+const data = Buffer.from('hello')
+const cid = new CID('QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n')
+const block = new Block(data, cid)
+
+const { port1, port2 } = new MessageChannel()
+
+// Will copy underyling memory
+port1.postMessage(encodeBlock(block))
+
+// Will transfer underlying memory (block & cid will be corrupt on this thread)
+const transfer = []
+port1.postMessage(encodeBlock(block, transfer), transfer)
+
+
+// On the receiver thread
+port2.onmessage = ({data}) => {
+  const block = decodeBlock(data)
+  block instanceof Block // true
+}
+```
+
+### DAGNode
+
+Codec for DAGNodes accepted by `ipfs.dag.put` API.
+
+```js
+const { encodeNode, decodeNode } = require('ipfs-message-port-protocol/src/dag')
+
+
+const cid = CID('QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n')
+const dagNode = { hi: 'hello', link: cid }
+
+const { port1, port2 } = new MessageChannel()
+
+// Will copy underyling memory
+port1.postMessage(encodeNode(dagNode))
+
+// Will transfer underlying memory (`dagNode.link` will be corrupt on this thread)
+const transfer = []
+port1.postMessage(encodeNode(dagNode, transfer), transfer)
+
+
+// On the receiver thread
+port2.onmessage = ({data}) => {
+  const dagNode = decodeNode(data)
+  dagNode.link instanceof CID // true
+}
+```
+
+### AsyncIterable
+
+Encoder allows producer to encode [async iterables][] such that it can be transferred across threads and decoded by a consumer on the other end and take care of all the IO coordination between two. Unlike other encoders `transfer` argument is mandatory (because value is encoded to a [MessagePort][] that can only be transferred). Additionally encoder / decoder take item encoder / decoder functions to encode each item of the async iterable.
```

Comment:
```suggestion
This encoder encodes [async iterables][] such that they can be transferred across threads and decoded by a consumer on the other end while taking care of all the IO coordination between two. Unlike other encoders the `transfer` argument is mandatory (because value is encoded to a [MessagePort][] that can only be transferred). Additionally encoder / decoder can take item encoder / decoder functions to encode each item of the async iterable.
```

Reply:
If encoder/decoder functions are optional, please put them in an options argument.

Reply:
they aren't optional.

Reply:
Changed wording to be more clear about that.